In [215]:
import pandas as pd

### Data dictionary, copied from Stathead.

Rk -- Rank
HR -- Home Runs
From -- From Season
To -- To Season
G -- Games Played
PA -- Plate Appearances
When available, we use actual plate appearances from play-by-play game accounts
Otherwise estimated using AB + BB + HBP + SF + SH,
which excludes catcher interferences.
When this color click for a summary of each PA.
AB -- At Bats
R -- Runs Scored
H -- Hits
1B -- Singles
2B -- Doubles
3B -- Triples
HR -- Home Runs
RBI -- Runs Batted In
SB -- Stolen Bases
CS -- Caught Stealing
BB -- Bases on Balls/Walks
SO -- Strikeouts
BA -- Hits/At Bats
For recent years, leaders need 3.1 PA
per team game played
Bold indicates highest BA using current stats
Gold means awarded title at end of year.
OBP -- (H + BB + HBP)/(At Bats + BB + HBP + SF)
For recent years, leaders need 3.1 PA
per team game played
SLG -- Total Bases/At Bats or
(1B + 2*2B + 3*3B + 4*HR)/AB
For recent years, leaders need 3.1 PA
per team game played
OPS -- On-Base + Slugging Percentages
For recent years, leaders need 3.1 PA
per team game played
OPS+ -- OPS+
100*[OBP/lg OBP + SLG/lg SLG - 1]
Adjusted to the player’s ballpark(s)
TB -- Total Bases
Singles + 2 x Doubles + 3 x Triples + 4 x Home Runs.
GIDP -- Double Plays Grounded Into
Only includes standard 6-4-3, 4-3, etc. double plays.
First tracked in 1933.
For gamelogs only in seasons we have play-by-play, we include triple plays as well.
All official seasonal totals do not include GITP's.
HBP -- Times Hit by a Pitch
SH -- Sacrifice Hits (Sacrifice Bunts)
SF -- Sacrifice Flies
First tracked in 1955.
IBB -- Intentional Bases on Balls
First tracked in 1955.
Pos -- Positions Played
The positions either followed by the games played at that position
or in order of games or innings played.
For a single season, * indicates they played at least 2/3rds of the team games there
Positions after / indicate less than ten games played at those positions.
For career, a + sign means more than 300 games at that position and
a - sign means less than 30 games.

### Import the dataset, find data types and nulls

In [216]:
players = pd.read_csv("/Users/matthewcavanaugh/Desktop/IOD Data camp/Project 3/100 HR ret 2018 - Sheet0.csv")

In [217]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 803 entries, 0 to 802
Data columns (total 34 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 803 non-null    int64  
 1   Player             803 non-null    object 
 2   HR                 803 non-null    int64  
 3   From               803 non-null    int64  
 4   To                 803 non-null    int64  
 5   Age                803 non-null    object 
 6   G                  803 non-null    int64  
 7   PA                 803 non-null    int64  
 8   AB                 803 non-null    int64  
 9   R                  803 non-null    int64  
 10  H                  803 non-null    int64  
 11  1B                 803 non-null    int64  
 12  2B                 803 non-null    int64  
 13  3B                 803 non-null    int64  
 14  HR.1               803 non-null    int64  
 15  RBI                803 non-null    int64  
 16  SB                 803 non

In [218]:
players.isnull().sum()

Rk                    0
Player                0
HR                    0
From                  0
To                    0
Age                   0
G                     0
PA                    0
AB                    0
R                     0
H                     0
1B                    0
2B                    0
3B                    0
HR.1                  0
RBI                   0
SB                    0
CS                   13
BB                    0
SO                    5
BA                    0
OBP                   0
SLG                   0
OPS                   0
OPS+                  0
TB                    0
GIDP                 26
HBP                   0
SH                    1
SF                   96
IBB                  11
Pos                   0
Team                  0
Player-additional     0
dtype: int64

### Fill nulls with 0

In [219]:
players.fillna(0, inplace=True)
players

Rk          Player   HR  From    To    Age     G     PA     AB     R  \
0      1     Barry Bonds  762  1986  2007  21-42  2986  12606   9847  2227   
1      2     Henry Aaron  755  1954  1976  20-42  3298  13941  12364  2174   
2      3       Babe Ruth  714  1914  1935  19-40  2503  10628   8399  2174   
3      4  Álex Rodríguez  696  1994  2016  18-40  2784  12207  10566  2021   
4      5     Willie Mays  660  1948  1973  17-42  3005  12545  10924  2068   
..   ...             ...  ...   ...   ...    ...   ...    ...    ...   ...   
798  799     Augie Galan  100  1934  1949  22-37  1742   7007   5937  1004   
799  800     Mike Jacobs  100  2005  2012  24-31   569   2140   1949   248   
800  801       John Kruk  100  1986  1995  25-34  1200   4603   3897   582   
801  802       Duke Sims  100  1964  1974  23-33   843   2810   2422   263   
802  803   Randy Velarde  100  1987  2002  24-39  1273   4813   4244   633   

     ...  OPS+    TB   GIDP  HBP     SH     SF    IBB         Pos  \
0    ...   182  5976  165.0  106    4.0   91.0  688.0     *78HD/9   
1    ...   155  6856  328.0   32   21.0  121.0  293.0  *983D7H4/5   
2    ...   206  5793    2.0   43  114.0    0.0   79.0      791H83   
3    ...   140  5813  261.0  176   16.0  111.0   97.0      56DH/3   
4    ...   155  6080  251.0   45   13.0   91.0  214.0    *8H37/96   
..   ...   ...   ...    ...  ...    ...    ...    ...         ...   
798  ...   122  2490   72.0   25   62.0    0.0   45.0   78H3549/6   
799  ...   104   921   49.0    7    1.0   13.0   18.0         3DH   
800  ...   134  1737   92.0    2   12.0   43.0   83.0      37H9D8   
801  ...   112   972   45.0   35    6.0    9.0   34.0     2H379/D   
802  ...   101  1731  118.0   49   38.0   19.0    6.0  4567HD39/8   

                Team  Player-additional  
0             PITSFG          bondsba01  
1          ATLMILMLN          aaronha01  
2          BOSBSNNYY           ruthba01  
3          NYYSEATEX          rodrial01  
4       BBBNYGNYMSFG           mayswi01  
..               ...                ...  
798  BROCHCCINNYGPHA          galanau01  
799     ARIFLAKCRNYM          jacobmi02  
800        CHWPHISDP           krukjo01  
801  CLEDETLADNYYTEX           simsdu01  
802  ANACALNYYOAKTEX          velarra01  

[803 rows x 34 columns]

### We're going to create new columns that separately list a player's debut and retirement ages. First we will remove the dash so that "Age" contains two separate strings. Then we will extract the first and second strings to create new columns called "Debut_age" and "Ret_age" respectively.

In [220]:
players['Age'] = players['Age'].str.replace('-', ' ')
players

Rk          Player   HR  From    To    Age     G     PA     AB     R  \
0      1     Barry Bonds  762  1986  2007  21 42  2986  12606   9847  2227   
1      2     Henry Aaron  755  1954  1976  20 42  3298  13941  12364  2174   
2      3       Babe Ruth  714  1914  1935  19 40  2503  10628   8399  2174   
3      4  Álex Rodríguez  696  1994  2016  18 40  2784  12207  10566  2021   
4      5     Willie Mays  660  1948  1973  17 42  3005  12545  10924  2068   
..   ...             ...  ...   ...   ...    ...   ...    ...    ...   ...   
798  799     Augie Galan  100  1934  1949  22 37  1742   7007   5937  1004   
799  800     Mike Jacobs  100  2005  2012  24 31   569   2140   1949   248   
800  801       John Kruk  100  1986  1995  25 34  1200   4603   3897   582   
801  802       Duke Sims  100  1964  1974  23 33   843   2810   2422   263   
802  803   Randy Velarde  100  1987  2002  24 39  1273   4813   4244   633   

     ...  OPS+    TB   GIDP  HBP     SH     SF    IBB         Pos  \
0    ...   182  5976  165.0  106    4.0   91.0  688.0     *78HD/9   
1    ...   155  6856  328.0   32   21.0  121.0  293.0  *983D7H4/5   
2    ...   206  5793    2.0   43  114.0    0.0   79.0      791H83   
3    ...   140  5813  261.0  176   16.0  111.0   97.0      56DH/3   
4    ...   155  6080  251.0   45   13.0   91.0  214.0    *8H37/96   
..   ...   ...   ...    ...  ...    ...    ...    ...         ...   
798  ...   122  2490   72.0   25   62.0    0.0   45.0   78H3549/6   
799  ...   104   921   49.0    7    1.0   13.0   18.0         3DH   
800  ...   134  1737   92.0    2   12.0   43.0   83.0      37H9D8   
801  ...   112   972   45.0   35    6.0    9.0   34.0     2H379/D   
802  ...   101  1731  118.0   49   38.0   19.0    6.0  4567HD39/8   

                Team  Player-additional  
0             PITSFG          bondsba01  
1          ATLMILMLN          aaronha01  
2          BOSBSNNYY           ruthba01  
3          NYYSEATEX          rodrial01  
4       BBBNYGNYMSFG           mayswi01  
..               ...                ...  
798  BROCHCCINNYGPHA          galanau01  
799     ARIFLAKCRNYM          jacobmi02  
800        CHWPHISDP           krukjo01  
801  CLEDETLADNYYTEX           simsdu01  
802  ANACALNYYOAKTEX          velarra01  

[803 rows x 34 columns]

In [221]:
players['Debut_age'] = players['Age'].str.extract('(\d+)')

<>:1: SyntaxWarning: invalid escape sequence '\d'
<>:1: SyntaxWarning: invalid escape sequence '\d'
/var/folders/cf/4zrkk0lx1nbdx9tpklpk4jc00000gn/T/ipykernel_58288/1073748372.py:1: SyntaxWarning: invalid escape sequence '\d'
  players['Debut_age'] = players['Age'].str.extract('(\d+)')


In [222]:
players['Debut_age'].value_counts()

Debut_age
22    186
23    142
21    141
24    101
20     85
19     55
25     38
26     19
18     17
27      8
17      7
15      1
29      1
31      1
28      1
Name: count, dtype: int64

In [223]:
players['Ret_age'] = players['Age'].str.extract(r'(\d+)\s+(\d+)').iloc[:, 1]
players

Rk          Player   HR  From    To    Age     G     PA     AB     R  \
0      1     Barry Bonds  762  1986  2007  21 42  2986  12606   9847  2227   
1      2     Henry Aaron  755  1954  1976  20 42  3298  13941  12364  2174   
2      3       Babe Ruth  714  1914  1935  19 40  2503  10628   8399  2174   
3      4  Álex Rodríguez  696  1994  2016  18 40  2784  12207  10566  2021   
4      5     Willie Mays  660  1948  1973  17 42  3005  12545  10924  2068   
..   ...             ...  ...   ...   ...    ...   ...    ...    ...   ...   
798  799     Augie Galan  100  1934  1949  22 37  1742   7007   5937  1004   
799  800     Mike Jacobs  100  2005  2012  24 31   569   2140   1949   248   
800  801       John Kruk  100  1986  1995  25 34  1200   4603   3897   582   
801  802       Duke Sims  100  1964  1974  23 33   843   2810   2422   263   
802  803   Randy Velarde  100  1987  2002  24 39  1273   4813   4244   633   

     ...   GIDP  HBP     SH     SF    IBB         Pos             Team  \
0    ...  165.0  106    4.0   91.0  688.0     *78HD/9           PITSFG   
1    ...  328.0   32   21.0  121.0  293.0  *983D7H4/5        ATLMILMLN   
2    ...    2.0   43  114.0    0.0   79.0      791H83        BOSBSNNYY   
3    ...  261.0  176   16.0  111.0   97.0      56DH/3        NYYSEATEX   
4    ...  251.0   45   13.0   91.0  214.0    *8H37/96     BBBNYGNYMSFG   
..   ...    ...  ...    ...    ...    ...         ...              ...   
798  ...   72.0   25   62.0    0.0   45.0   78H3549/6  BROCHCCINNYGPHA   
799  ...   49.0    7    1.0   13.0   18.0         3DH     ARIFLAKCRNYM   
800  ...   92.0    2   12.0   43.0   83.0      37H9D8        CHWPHISDP   
801  ...   45.0   35    6.0    9.0   34.0     2H379/D  CLEDETLADNYYTEX   
802  ...  118.0   49   38.0   19.0    6.0  4567HD39/8  ANACALNYYOAKTEX   

     Player-additional  Debut_age  Ret_age  
0            bondsba01         21       42  
1            aaronha01         20       42  
2             ruthba01         19       40  
3            rodrial01         18       40  
4             mayswi01         17       42  
..                 ...        ...      ...  
798          galanau01         22       37  
799          jacobmi02         24       31  
800           krukjo01         25       34  
801           simsdu01         23       33  
802          velarra01         24       39  

[803 rows x 36 columns]

### I've separately prepared a second dataframe listing whether each player in this dataset is a Hall of Famer. "Player-additional" is only the common column between the two, so we will join it onto the main database left, and name the new dataframe as "mlb".

In [224]:
hof = pd.read_csv("/Users/matthewcavanaugh/Desktop/IOD Data camp/Project 3/100 HR ret 2018 HOF v2 - Sheet0.csv")
hof

Player-additional  HOF
0           aaronha01    1
1            ruthba01    1
2            mayswi01    1
3           griffke02    1
4           thomeji01    1
..                ...  ...
122         friscfr01    1
123         lindsfr01    1
124         sislege01    1
125         delahed01    1
126         wagneho01    1

[127 rows x 2 columns]

In [225]:
mlb = players.merge(hof, on='Player-additional', how='left')
mlb

Rk          Player   HR  From    To    Age     G     PA     AB     R  \
0      1     Barry Bonds  762  1986  2007  21 42  2986  12606   9847  2227   
1      2     Henry Aaron  755  1954  1976  20 42  3298  13941  12364  2174   
2      3       Babe Ruth  714  1914  1935  19 40  2503  10628   8399  2174   
3      4  Álex Rodríguez  696  1994  2016  18 40  2784  12207  10566  2021   
4      5     Willie Mays  660  1948  1973  17 42  3005  12545  10924  2068   
..   ...             ...  ...   ...   ...    ...   ...    ...    ...   ...   
798  799     Augie Galan  100  1934  1949  22 37  1742   7007   5937  1004   
799  800     Mike Jacobs  100  2005  2012  24 31   569   2140   1949   248   
800  801       John Kruk  100  1986  1995  25 34  1200   4603   3897   582   
801  802       Duke Sims  100  1964  1974  23 33   843   2810   2422   263   
802  803   Randy Velarde  100  1987  2002  24 39  1273   4813   4244   633   

     ...  HBP     SH     SF    IBB         Pos             Team  \
0    ...  106    4.0   91.0  688.0     *78HD/9           PITSFG   
1    ...   32   21.0  121.0  293.0  *983D7H4/5        ATLMILMLN   
2    ...   43  114.0    0.0   79.0      791H83        BOSBSNNYY   
3    ...  176   16.0  111.0   97.0      56DH/3        NYYSEATEX   
4    ...   45   13.0   91.0  214.0    *8H37/96     BBBNYGNYMSFG   
..   ...  ...    ...    ...    ...         ...              ...   
798  ...   25   62.0    0.0   45.0   78H3549/6  BROCHCCINNYGPHA   
799  ...    7    1.0   13.0   18.0         3DH     ARIFLAKCRNYM   
800  ...    2   12.0   43.0   83.0      37H9D8        CHWPHISDP   
801  ...   35    6.0    9.0   34.0     2H379/D  CLEDETLADNYYTEX   
802  ...   49   38.0   19.0    6.0  4567HD39/8  ANACALNYYOAKTEX   

     Player-additional  Debut_age  Ret_age  HOF  
0            bondsba01         21       42  NaN  
1            aaronha01         20       42  1.0  
2             ruthba01         19       40  1.0  
3            rodrial01         18       40  NaN  
4             mayswi01         17       42  1.0  
..                 ...        ...      ...  ...  
798          galanau01         22       37  NaN  
799          jacobmi02         24       31  NaN  
800           krukjo01         25       34  NaN  
801           simsdu01         23       33  NaN  
802          velarra01         24       39  NaN  

[803 rows x 37 columns]

### Non-Hall of Famers now have nulls in the "HOF" column, so let's fill those with 0s.

In [226]:
mlb.fillna(0, inplace=True)

In [227]:
mlb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 803 entries, 0 to 802
Data columns (total 37 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 803 non-null    int64  
 1   Player             803 non-null    object 
 2   HR                 803 non-null    int64  
 3   From               803 non-null    int64  
 4   To                 803 non-null    int64  
 5   Age                803 non-null    object 
 6   G                  803 non-null    int64  
 7   PA                 803 non-null    int64  
 8   AB                 803 non-null    int64  
 9   R                  803 non-null    int64  
 10  H                  803 non-null    int64  
 11  1B                 803 non-null    int64  
 12  2B                 803 non-null    int64  
 13  3B                 803 non-null    int64  
 14  HR.1               803 non-null    int64  
 15  RBI                803 non-null    int64  
 16  SB                 803 non

In [228]:
mlb.drop('Age', axis=1, inplace=True)

In [229]:
mlb['HOF'] = mlb['HOF'].astype(int)

In [230]:
mlb

Rk          Player   HR  From    To     G     PA     AB     R     H  \
0      1     Barry Bonds  762  1986  2007  2986  12606   9847  2227  2935   
1      2     Henry Aaron  755  1954  1976  3298  13941  12364  2174  3771   
2      3       Babe Ruth  714  1914  1935  2503  10628   8399  2174  2873   
3      4  Álex Rodríguez  696  1994  2016  2784  12207  10566  2021  3115   
4      5     Willie Mays  660  1948  1973  3005  12545  10924  2068  3293   
..   ...             ...  ...   ...   ...   ...    ...    ...   ...   ...   
798  799     Augie Galan  100  1934  1949  1742   7007   5937  1004  1706   
799  800     Mike Jacobs  100  2005  2012   569   2140   1949   248   493   
800  801       John Kruk  100  1986  1995  1200   4603   3897   582  1170   
801  802       Duke Sims  100  1964  1974   843   2810   2422   263   580   
802  803   Randy Velarde  100  1987  2002  1273   4813   4244   633  1171   

     ...  HBP     SH     SF    IBB         Pos             Team  \
0    ...  106    4.0   91.0  688.0     *78HD/9           PITSFG   
1    ...   32   21.0  121.0  293.0  *983D7H4/5        ATLMILMLN   
2    ...   43  114.0    0.0   79.0      791H83        BOSBSNNYY   
3    ...  176   16.0  111.0   97.0      56DH/3        NYYSEATEX   
4    ...   45   13.0   91.0  214.0    *8H37/96     BBBNYGNYMSFG   
..   ...  ...    ...    ...    ...         ...              ...   
798  ...   25   62.0    0.0   45.0   78H3549/6  BROCHCCINNYGPHA   
799  ...    7    1.0   13.0   18.0         3DH     ARIFLAKCRNYM   
800  ...    2   12.0   43.0   83.0      37H9D8        CHWPHISDP   
801  ...   35    6.0    9.0   34.0     2H379/D  CLEDETLADNYYTEX   
802  ...   49   38.0   19.0    6.0  4567HD39/8  ANACALNYYOAKTEX   

     Player-additional  Debut_age  Ret_age  HOF  
0            bondsba01         21       42    0  
1            aaronha01         20       42    1  
2             ruthba01         19       40    1  
3            rodrial01         18       40    0  
4             mayswi01         17       42    1  
..                 ...        ...      ...  ...  
798          galanau01         22       37    0  
799          jacobmi02         24       31    0  
800           krukjo01         25       34    0  
801           simsdu01         23       33    0  
802          velarra01         24       39    0  

[803 rows x 36 columns]

### Now we need to make sure there are no duplicates. This is where "Player-additional" comes into play - it is a unique identifier to avoid confusion between players with identical names.

In [231]:
mlb['Player-additional'].nunique()

803

In [232]:
mlb['Player'].nunique()

800

### There are three doppelgangers! Let's run value_counts to see which names appear more thant wice.

In [233]:
mlb['Player'].value_counts()

Player
José Cruz          2
Gary Matthews      2
Frank Thomas       2
Barry Bonds        1
Aaron Rowand       1
                  ..
Bob Horner         1
Benito Santiago    1
Ron Fairly         1
Pedro Guerrero     1
Randy Velarde      1
Name: count, Length: 800, dtype: int64

### Now let's pull up their indexes and rename some players to avoid confusion. Jose Cruz and Gary Matthews are both father-son pairs, so we will rename the younger ones with Jr. Among the Frank Thomases, one was primarily an outfield and the other a designated hatter, so we will add those positions in parentheses.

In [234]:
mlb[mlb['Player'] == 'José Cruz']

Rk     Player   HR  From    To     G    PA    AB     R     H  ...  HBP  \
299  300  José Cruz  204  1997  2008  1388  5448  4724   713  1167  ...    5   
410  411  José Cruz  165  1970  1988  2353  8931  7917  1036  2251  ...    7   

       SH    SF    IBB       Pos                         Team  \
299  23.0  38.0   38.0   *897H/D  ARIBOSHOULADSDPSEASFGTBDTOR   
410  27.0  82.0  142.0  *798HD/3                    HOUNYYSTL   

     Player-additional  Debut_age  Ret_age  HOF  
299           cruzjo02         23       34    0  
410           cruzjo01         22       40    0  

[2 rows x 36 columns]

In [235]:
mlb.loc[410, 'Player'] = mlb.loc[410, 'Player'].replace('José Cruz', 'José Cruz Jr.')

In [236]:
mlb[mlb['Player'] == 'Gary Matthews']

Rk         Player   HR  From    To     G    PA    AB     R     H  ...  \
246  247  Gary Matthews  234  1972  1987  2033  8189  7147  1083  2011  ...   
719  720  Gary Matthews  108  1999  2010  1281  4617  4103   612  1056  ...   

     HBP    SH    SF   IBB    Pos                   Team  Player-additional  \
246   21  19.0  62.0  46.0  *79HD        ATLCHCPHISEASFG          matthga01   
719   18  13.0  29.0  26.0  897HD  BALCHCLAANYMPITSDPTEX          matthga02   

     Debut_age  Ret_age  HOF  
246         21       36    0  
719         24       35    0  

[2 rows x 36 columns]

In [237]:
mlb.loc[719, 'Player'] = mlb.loc[719, 'Player'].replace('Gary Matthews', 'Gary Matthews Jr.')

In [238]:
mlb[mlb['Player'] == 'Frank Thomas']

Rk        Player   HR  From    To     G     PA    AB     R     H  ...  \
19    20  Frank Thomas  521  1990  2008  2322  10075  8199  1494  2468  ...   
157  158  Frank Thomas  286  1951  1966  1766   6917  6285   792  1671  ...   

     HBP    SH     SF    IBB       Pos                   Team  \
19    87   0.0  121.0  168.0       D3H              CHWOAKTOR   
157   51  26.0   70.0   64.0  7583H9/4  CHCCINHOUMLNNYMPHIPIT   

     Player-additional  Debut_age  Ret_age  HOF  
19           thomafr04         22       40    1  
157          thomafr03         22       37    0  

[2 rows x 36 columns]

In [239]:
mlb.loc[19, 'Player'] = mlb.loc[19, 'Player'].replace('Frank Thomas', 'Frank Thomas (DH)')
mlb.loc[157, 'Player'] = mlb.loc[157, 'Player'].replace('Frank Thomas', 'Frank Thomas (OF)')

### Check nunique in "Player" again to ensure the duplicates have been separated. There should be 803.

In [240]:
mlb['Player'].nunique()

803

### Success!

In [241]:
mlb.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 803 entries, 0 to 802
Data columns (total 36 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Rk                 803 non-null    int64  
 1   Player             803 non-null    object 
 2   HR                 803 non-null    int64  
 3   From               803 non-null    int64  
 4   To                 803 non-null    int64  
 5   G                  803 non-null    int64  
 6   PA                 803 non-null    int64  
 7   AB                 803 non-null    int64  
 8   R                  803 non-null    int64  
 9   H                  803 non-null    int64  
 10  1B                 803 non-null    int64  
 11  2B                 803 non-null    int64  
 12  3B                 803 non-null    int64  
 13  HR.1               803 non-null    int64  
 14  RBI                803 non-null    int64  
 15  SB                 803 non-null    int64  
 16  CS                 803 non

### Let's drop a few more columns we no longer need:


In [242]:
mlb.drop('Player-additional', axis=1, inplace=True)
mlb.drop('Rk', axis=1, inplace=True)
mlb.drop('Team', axis=1, inplace=True)
mlb.drop('Pos', axis=1, inplace=True)
mlb.drop('HR.1', axis=1, inplace=True)

### A couple columns are still non-numeric. Let's fix that.

In [243]:
mlb['Debut_age'] = mlb['Debut_age'].astype(int)
mlb['Ret_age'] = mlb['Ret_age'].astype(int)

### For appearances' sake, let's rename and rearrange the order of some columns.

In [244]:
mlb.rename(columns={'From': 'Debuted'}, inplace=True)
mlb.rename(columns={'To': 'Retired'}, inplace=True)
mlb

Player   HR  Debuted  Retired     G     PA     AB     R     H  \
0       Barry Bonds  762     1986     2007  2986  12606   9847  2227  2935   
1       Henry Aaron  755     1954     1976  3298  13941  12364  2174  3771   
2         Babe Ruth  714     1914     1935  2503  10628   8399  2174  2873   
3    Álex Rodríguez  696     1994     2016  2784  12207  10566  2021  3115   
4       Willie Mays  660     1948     1973  3005  12545  10924  2068  3293   
..              ...  ...      ...      ...   ...    ...    ...   ...   ...   
798     Augie Galan  100     1934     1949  1742   7007   5937  1004  1706   
799     Mike Jacobs  100     2005     2012   569   2140   1949   248   493   
800       John Kruk  100     1986     1995  1200   4603   3897   582  1170   
801       Duke Sims  100     1964     1974   843   2810   2422   263   580   
802   Randy Velarde  100     1987     2002  1273   4813   4244   633  1171   

       1B  ...  OPS+    TB   GIDP  HBP     SH     SF    IBB  Debut_age  \
0    1495  ...   182  5976  165.0  106    4.0   91.0  688.0         21   
1    2294  ...   155  6856  328.0   32   21.0  121.0  293.0         20   
2    1517  ...   206  5793    2.0   43  114.0    0.0   79.0         19   
3    1840  ...   140  5813  261.0  176   16.0  111.0   97.0         18   
4    1967  ...   155  6080  251.0   45   13.0   91.0  214.0         17   
..    ...  ...   ...   ...    ...  ...    ...    ...    ...        ...   
798  1196  ...   122  2490   72.0   25   62.0    0.0   45.0         22   
799   271  ...   104   921   49.0    7    1.0   13.0   18.0         24   
800   837  ...   134  1737   92.0    2   12.0   43.0   83.0         25   
801   394  ...   112   972   45.0   35    6.0    9.0   34.0         23   
802   834  ...   101  1731  118.0   49   38.0   19.0    6.0         24   

     Ret_age  HOF  
0         42    0  
1         42    1  
2         40    1  
3         40    0  
4         42    1  
..       ...  ...  
798       37    0  
799       31    0  
800       34    0  
801       33    0  
802       39    0  

[803 rows x 31 columns]

In [245]:
mlb = mlb[[
 'Player',
 'Debuted',
 'Debut_age',
 'Retired',
 'Ret_age',
 'G',
 'PA',
 'AB',
 'R',
 'H',
 '1B',
 '2B',
 '3B',
 'HR',
 'RBI',
 'SB',
 'CS',
 'BB',
 'SO',
 'BA',
 'OBP',
 'SLG',
 'OPS',
 'OPS+',
 'TB',
 'GIDP',
 'HBP',
 'SH',
 'SF',
 'IBB',
 'HOF']]

### Set 'player' as index

In [246]:
mlb.set_index('Player', inplace=True)
mlb

Debuted  Debut_age  Retired  Ret_age     G     PA     AB  \
Player                                                                     
Barry Bonds        1986         21     2007       42  2986  12606   9847   
Henry Aaron        1954         20     1976       42  3298  13941  12364   
Babe Ruth          1914         19     1935       40  2503  10628   8399   
Álex Rodríguez     1994         18     2016       40  2784  12207  10566   
Willie Mays        1948         17     1973       42  3005  12545  10924   
...                 ...        ...      ...      ...   ...    ...    ...   
Augie Galan        1934         22     1949       37  1742   7007   5937   
Mike Jacobs        2005         24     2012       31   569   2140   1949   
John Kruk          1986         25     1995       34  1200   4603   3897   
Duke Sims          1964         23     1974       33   843   2810   2422   
Randy Velarde      1987         24     2002       39  1273   4813   4244   

                   R     H    1B  ...    SLG    OPS  OPS+    TB   GIDP  HBP  \
Player                            ...                                         
Barry Bonds     2227  2935  1495  ...  0.607  1.051   182  5976  165.0  106   
Henry Aaron     2174  3771  2294  ...  0.555  0.928   155  6856  328.0   32   
Babe Ruth       2174  2873  1517  ...  0.690  1.164   206  5793    2.0   43   
Álex Rodríguez  2021  3115  1840  ...  0.550  0.930   140  5813  261.0  176   
Willie Mays     2068  3293  1967  ...  0.557  0.940   155  6080  251.0   45   
...              ...   ...   ...  ...    ...    ...   ...   ...    ...  ...   
Augie Galan     1004  1706  1196  ...  0.419  0.810   122  2490   72.0   25   
Mike Jacobs      248   493   271  ...  0.473  0.786   104   921   49.0    7   
John Kruk        582  1170   837  ...  0.446  0.842   134  1737   92.0    2   
Duke Sims        263   580   394  ...  0.401  0.741   112   972   45.0   35   
Randy Velarde    633  1171   834  ...  0.408  0.760   101  1731  118.0   49   

                   SH     SF    IBB  HOF  
Player                                    
Barry Bonds       4.0   91.0  688.0    0  
Henry Aaron      21.0  121.0  293.0    1  
Babe Ruth       114.0    0.0   79.0    1  
Álex Rodríguez   16.0  111.0   97.0    0  
Willie Mays      13.0   91.0  214.0    1  
...               ...    ...    ...  ...  
Augie Galan      62.0    0.0   45.0    0  
Mike Jacobs       1.0   13.0   18.0    0  
John Kruk        12.0   43.0   83.0    0  
Duke Sims         6.0    9.0   34.0    0  
Randy Velarde    38.0   19.0    6.0    0  

[803 rows x 30 columns]

### Remove accent marks from player names. This is necessary for a function we will use in the main lab.

In [254]:
import unicodedata

In [255]:
def remove_accents_from_index(df):
    """
    Removes accents from the index of the given dataframe.
    
    Parameters:
    df (pd.DataFrame): The dataframe whose index needs to be cleaned.
    
    Returns:
    pd.DataFrame: A new dataframe with accents removed from the index.
    """
    # Function to remove accents
    def remove_accents(text):
        return ''.join(
            c for c in unicodedata.normalize('NFD', text) 
            if unicodedata.category(c) != 'Mn'
        )
    
    # Apply the function to the index
    df.index = df.index.map(remove_accents)
    
    return df

In [256]:
mlb = remove_accents_from_index(mlb)

In [257]:
mlb

Debuted  Debut_age  Retired  Ret_age     G     PA     AB  \
Player                                                                     
Barry Bonds        1986         21     2007       42  2986  12606   9847   
Henry Aaron        1954         20     1976       42  3298  13941  12364   
Babe Ruth          1914         19     1935       40  2503  10628   8399   
Alex Rodriguez     1994         18     2016       40  2784  12207  10566   
Willie Mays        1948         17     1973       42  3005  12545  10924   
...                 ...        ...      ...      ...   ...    ...    ...   
Augie Galan        1934         22     1949       37  1742   7007   5937   
Mike Jacobs        2005         24     2012       31   569   2140   1949   
John Kruk          1986         25     1995       34  1200   4603   3897   
Duke Sims          1964         23     1974       33   843   2810   2422   
Randy Velarde      1987         24     2002       39  1273   4813   4244   

                   R     H    1B  ...    SLG    OPS  OPS+    TB   GIDP  HBP  \
Player                            ...                                         
Barry Bonds     2227  2935  1495  ...  0.607  1.051   182  5976  165.0  106   
Henry Aaron     2174  3771  2294  ...  0.555  0.928   155  6856  328.0   32   
Babe Ruth       2174  2873  1517  ...  0.690  1.164   206  5793    2.0   43   
Alex Rodriguez  2021  3115  1840  ...  0.550  0.930   140  5813  261.0  176   
Willie Mays     2068  3293  1967  ...  0.557  0.940   155  6080  251.0   45   
...              ...   ...   ...  ...    ...    ...   ...   ...    ...  ...   
Augie Galan     1004  1706  1196  ...  0.419  0.810   122  2490   72.0   25   
Mike Jacobs      248   493   271  ...  0.473  0.786   104   921   49.0    7   
John Kruk        582  1170   837  ...  0.446  0.842   134  1737   92.0    2   
Duke Sims        263   580   394  ...  0.401  0.741   112   972   45.0   35   
Randy Velarde    633  1171   834  ...  0.408  0.760   101  1731  118.0   49   

                   SH     SF    IBB  HOF  
Player                                    
Barry Bonds       4.0   91.0  688.0    0  
Henry Aaron      21.0  121.0  293.0    1  
Babe Ruth       114.0    0.0   79.0    1  
Alex Rodriguez   16.0  111.0   97.0    0  
Willie Mays      13.0   91.0  214.0    1  
...               ...    ...    ...  ...  
Augie Galan      62.0    0.0   45.0    0  
Mike Jacobs       1.0   13.0   18.0    0  
John Kruk        12.0   43.0   83.0    0  
Duke Sims         6.0    9.0   34.0    0  
Randy Velarde    38.0   19.0    6.0    0  

[803 rows x 30 columns]

### Now our dataset is clean and ready to be saved!

In [258]:
#mlb.to_csv('100 Home runs cleaned.csv')